lib/cli/commands/build.js

```diff
@@ -98,11 +125,57 @@ async function handleBuild(argv) {
 	}
 
 	const tree = await normalizer.generateProjectTree(normalizerOptions);
+	const buildSettings = (tree.builder && tree.builder.settings) || {};
+
+	const includedDependencies = buildHelper.createDependencyList({
+		tree: tree,
+		dependencies: argv["include-dependency"],
+		dependenciesRegExp: argv["include-dependency-regexp"]
+	});
+	const includedDependencyTrees = buildHelper.createDependencyList({
+		tree: tree,
+		dependencies: argv["include-dependency-tree"],
+		handleSubtree: true
+	});
+	const excludedDependencies = buildHelper.createDependencyList({
+		tree: tree,
+		dependencies: argv["exclude-dependency"],
+		dependenciesRegExp: argv["exclude-dependency-regexp"]
+	});
+	const excludedDependencyTrees = buildHelper.createDependencyList({
+		tree: tree,
+		dependencies: argv["exclude-dependency-tree"],
+		handleSubtree: true
+	});
+	const defaultIncludedDependencies = buildHelper.createDependencyList({
+		tree: tree,
+		dependencies: buildSettings.includeDependency,
+		dependenciesRegExp: buildSettings.includeDependencyRegExp
+	});
+	const defaultIncludedDependencyTrees = buildHelper.createDependencyList({
+		tree: tree,
+		dependencies: buildSettings.includeDependencyTree,
+		handleSubtree: true
+	});
+
+	// Append resolved include-trees to includedDependencies, ignoring excluded projects
+	buildHelper.mergeDependencyLists(includedDependencies, includedDependencyTrees, excludedDependencies);
+	// Append resolved exclude-trees to excludedDependencies, ignoring explicitly included projects
+	buildHelper.mergeDependencyLists(excludedDependencies, excludedDependencyTrees, includedDependencies);
+
+	// Combine dependencies set in build settings and append them to includedDependencies, ignoring excluded projects
+	Array.prototype.push.apply(defaultIncludedDependencies, defaultIncludedDependencyTrees);
+	buildHelper.mergeDependencyLists(includedDependencies, defaultIncludedDependencies, excludedDependencies);
+
+	const buildAll = buildHelper.alignWithBuilderApi(argv.all, includedDependencies, excludedDependencies);
+
 	await builder.build({
 		tree: tree,
 		destPath: argv.dest,
 		cleanDest: argv["clean-dest"],
-		buildDependencies: argv.all,
+		buildDependencies: buildAll,
```

Comment:
just out of curiosity: why does the builder still require include/exclude? We're already traversing the tree, looking at the new filtering options. Why can't we resolve the filters in the CLI layer and just give a list of projects (includes maybe) to the builder?

Reply:
> why does the builder still require include/exclude?

For consumers like Kristian, who use the Node.js API of the builder directly. I would try to keep the CLI layer as simple as possible and just pass the includes/excludes to the builder mostly untouched.